### PR TITLE
`Plagiarism checks`: Prevent multiple simultaneous plagiarism checks in the same course

### PR DESCRIPTION
**Pull request by Dominik-Weinzierl**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

@krusche asked me to disable simultaneous plagiarism checks per course to avoid performance issues and to avoid that users execute multiple plagiarism checks accidentally in case they are slow (e.g. in exercises with 2000 participating users) 

### Description

To prevent simultaneous plagiarism checks per course, we store a set of course ids that indicates whether a plagiarism check is currently active. This map is stored in our distributed Hazelcast cache and therefore available on all instances.

If the user tries to run multiple plagiarism checks at the same time, an exception is thrown.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Exercises of one Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to each exercise and the plagiarism check overview
4. Click both buttons of the plagiarism check at the same time
5. The second one should fail with an alert

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Screenshots

https://user-images.githubusercontent.com/60005702/175806871-d4261fea-a1e8-407d-b918-56367e7c9080.mp4

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ModelingPlagiarismDetectionService.java [hunk 1]
```diff
@@ -46,6 +50,12 @@ public ModelingPlagiarismDetectionService(PlagiarismWebsocketService plagiarismW
      */
     public ModelingPlagiarismResult checkPlagiarism(ModelingExercise exerciseWithParticipationsSubmissionsResults, double minimumSimilarity, int minimumModelSize,
             int minimumScore) {
+        var courseId = exerciseWithParticipationsSubmissionsResults.getCourseViaExerciseGroupOrCourseMember().getId();
+        if (plagiarismCacheService.isActivePlagiarismCheck(courseId)) {
+            return null;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/ModelingPlagiarismDetectionService.java:55]: we should throw an exception here

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ModelingPlagiarismDetectionService.java [hunk 2]
```diff
@@ -55,6 +65,8 @@ public ModelingPlagiarismResult checkPlagiarism(ModelingExercise exerciseWithPar
 
         result.setExercise(exerciseWithParticipationsSubmissionsResults);
 
+        plagiarismCacheService.disablePlagiarismCheck(courseId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/ModelingPlagiarismDetectionService.java:68]: we need to make sure that this is always invoked, even if an exception occurs, so please put the code above in a try catch statement, re-throw the exception and invoked `disablePlagiarismCheck` in the final clause

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCache.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.hazelcast.config.*;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.map.IMap;
+
+public class PlagiarismCache {
+
+    private final Logger logger = LoggerFactory.getLogger(PlagiarismCache.class);
+
+    private final IMap<Long, Boolean> activePlagiarismChecksPerCourse;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCache.java:18]: add documentation that Long means `courseId` here

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public boolean isActivePlagiarismCheck(Long courseId) {
+        return plagiarismCache.isActivePlagiarismCheck(courseId);
+    }
+
+    /**
+     * There is an active plagiarism check in this course. The entry is set to true.
+     * @param courseId current course
+     */
+    public void enablePlagiarismCheck(Long courseId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java:34]:
```suggestion
    public void setActivePlagiarismCheck(Long courseId) {
```

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        plagiarismCache.setActivePlagiarismCheck(courseId, true);
+    }
+
+    /**
+     * There is no active plagiarism check anymore. The entry is set to false.
+     * @param courseId current course
+     */
+    public void disablePlagiarismCheck(Long courseId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java:42]:
```suggestion
    public void setInactivePlagiarismCheck(Long courseId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java [hunk 6]
```diff
@@ -437,6 +437,11 @@ public ResponseEntity<ModelingPlagiarismResult> checkPlagiarism(@PathVariable lo
         long start = System.nanoTime();
         log.info("Start modelingPlagiarismDetectionService.checkPlagiarism for exercise {}", exerciseId);
         ModelingPlagiarismResult result = modelingPlagiarismDetectionService.checkPlagiarism(modelingExercise, similarityThreshold / 100, minimumSize, minimumScore);
+
+        if (result == null) {
+            throw new BadRequestAlertException("Only one active plagiarism check per course allowed", "PlagiarismCheck", "oneActivePlagiarismCheck");
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:442]: please throw this exception directly in the `checkPlagiarism` method

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java (hunk 4 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/plagiarism/cache/PlagiarismCacheService.java:34]: enable does not sound good in my opinion, same below for disable

File: src/main/webapp/i18n/de/error.json [hunk 7]
```diff
@@ -45,6 +45,7 @@
         "submissionPolicyAlreadyDisabled": "Die Abgaberichtlinie konnte nicht deaktiviert werden, da diese bereits deaktiviert ist.",
         "submissionPolicyToggleFailedPolicyNotExist": "Für diese Programmieraufgabe konnte die Abgaberichtlinie nicht (de)aktiviert werden, da die Programmieraufgabe keine Abgaberichtlinie hat. ",
         "submissionPolicyActiveNull": "Abgaberichtlinien müssen aktiviert oder deaktiviert sein. Dieser Status darf nicht null sein.",
-        "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!"
+        "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!",
+        "oneActivePlagiarismCheck": "Es wird bereits eine Plagiatskontrolle in diesem Kurs durchgeführt!"
```
Comment (krusche) [inline src/main/webapp/i18n/de/error.json:49]: Add something like "Bitte warte bis die aktuelle Plagiatskontrolle abgeschlossen ist, bevor du eine weitere startest"

File: src/main/webapp/i18n/en/error.json [hunk 8]
```diff
@@ -45,6 +45,7 @@
         "submissionPolicyAlreadyDisabled": "The submission policy could not be disabled, because it is already inactive.",
         "submissionPolicyToggleFailedPolicyNotExist": "The submission policy could not be toggled, because the programming exercise does not have a submission policy.",
         "submissionPolicyActiveNull": "Submission policies must be activated or deactivated. Activation cannot be null.",
-        "submissionPolicyUpdateFailedPolicyNotExist": "The submission policy could not be updated, because the programming exercise does not have a submission policy."
+        "submissionPolicyUpdateFailedPolicyNotExist": "The submission policy could not be updated, because the programming exercise does not have a submission policy.",
+        "oneActivePlagiarismCheck": "There is already a plagiarism check in this course!"
```
Comment (krusche) [inline src/main/webapp/i18n/en/error.json:49]: Add something like "Please wait until the current plagiarism check has finished until you start another one"

**Review by krusche — CHANGES_REQUESTED** [review]
see my inline feedback

**Review by ge65cer — DISMISSED** [review]
Tested on TS2. Could not start another plagiarism check under any circumstances when one is running. 👍

File: src/main/webapp/i18n/en/error.json [hunk 9]
```diff
@@ -45,6 +45,7 @@
         "submissionPolicyAlreadyDisabled": "The submission policy could not be disabled, because it is already inactive.",
         "submissionPolicyToggleFailedPolicyNotExist": "The submission policy could not be toggled, because the programming exercise does not have a submission policy.",
         "submissionPolicyActiveNull": "Submission policies must be activated or deactivated. Activation cannot be null.",
-        "submissionPolicyUpdateFailedPolicyNotExist": "The submission policy could not be updated, because the programming exercise does not have a submission policy."
+        "submissionPolicyUpdateFailedPolicyNotExist": "The submission policy could not be updated, because the programming exercise does not have a submission policy.",
+        "oneActivePlagiarismCheck": "Please wait until the current plagiarism check has finished until you start another one."
```
Comment (ge65cer) [inline src/main/webapp/i18n/en/error.json:49]: Scenario: Two users want to start a plagiarism check for different exercises in the same course. Then this error message would be a bit more descriptive I'd say.
```suggestion
        "oneActivePlagiarismCheck": "Please wait until a plagiarism check in this course has finished until you start another one."
```

File: src/main/webapp/i18n/de/error.json [hunk 10]
```diff
@@ -45,6 +45,7 @@
         "submissionPolicyAlreadyDisabled": "Die Abgaberichtlinie konnte nicht deaktiviert werden, da diese bereits deaktiviert ist.",
         "submissionPolicyToggleFailedPolicyNotExist": "Für diese Programmieraufgabe konnte die Abgaberichtlinie nicht (de)aktiviert werden, da die Programmieraufgabe keine Abgaberichtlinie hat. ",
         "submissionPolicyActiveNull": "Abgaberichtlinien müssen aktiviert oder deaktiviert sein. Dieser Status darf nicht null sein.",
-        "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!"
+        "submissionPolicyUpdateFailedPolicyNotExist": "Die Abgaberichtlinie konnte nicht gespeichert werden, da die Programmieraufgabe keine Abgaberichtlinie hat!",
+        "oneActivePlagiarismCheck": "Bitte warte bis die aktuelle Plagiatskontrolle abgeschlossen ist, bevor du eine weitere startest."
```
Comment (ge65cer) [inline src/main/webapp/i18n/de/error.json:49]:
```suggestion
        "oneActivePlagiarismCheck": "Bitte warte bis eine Plagiatskontrolle in diesem Kurs abgeschlossen ist, bevor du eine weitere startest."
```

**Review by ge65cer — DISMISSED** [review]

**Comment by Dominik-Weinzierl** [comment]
Tested on ts1 after refactoring - works still as expected :+1:

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 11]
```diff
@@ -225,6 +225,10 @@ public final class Constants {
 
     public static final int HAZELCAST_QUIZ_EXERCISE_CACHE_SERIALIZER_ID = 1;
 
+    public static final String HAZELCAST_PLAGIARISM_PREFIX = "plagiarism-";
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:228]: I guess we should stay consistent in this PR, but in the future / follow-up I think these are another set of constants that should rather be located in their own classes (PlagiarismCacheService, etc).

**Review by Strohgelaender — APPROVED** [review]
Code looks good, only one comment that could be resolved in a followup refactoring.

**Review by max-bergmann — APPROVED** [review]
Tested on TS1. Works as expected

**Review by krusche — APPROVED** [review]
Code looks good to me now 👍

**Pull request merged**